File: 프로그래머스/lv3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.cpp
#include <string>
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> rotate(vector<vector<int>>& key) {
    int n = key.size();
    vector<vector<int>> result(n, vector<int>(n, 0));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            int x = j;
            int y = n - i - 1;
            result[x][y] = key[i][j];
        }
    }
    return result;
}

bool check_open(vector<vector<int>>& key, vector<vector<int>>& lock, int n, int m, int count) {
    for (int x = 0; x < m + n - 1; ++x) {
        for (int y = 0; y < m + n - 1; ++y) {
            bool OK = true;
            int result = 0;

            for (int i = 0; i < n; ++i) {
                for (int j = 0; j < n; ++j) {
                    int hx = x - (n - 1 - i);
                    int hy = y - (n - 1 - j);
                    if (hx < 0 || hy < 0 || hx >= m || hy >= m) continue;
                    if (key[i][j] == 1 && lock[hx][hy] == 1) {
                        OK = false;
                        break;
                    } else if (key[i][j] == 1 && lock[hx][hy] == 0) {
                        result += 1;
                    }
                }
                if (!OK) break;
            }
            if (OK && result == count) {
                return true;
            }
        }
    }
    return false;
}

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    int holl_count = 0;
    int n = key.size();
    int m = lock.size();
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < m; ++j) {
            if (lock[i][j] == 0) {
                holl_count += 1;
            }
        }
    }

    bool answer = false;
    for (int i = 0; i < 4; ++i) {
        answer = check_open(key, lock, n, m, holl_count);
        if (answer) {
            break;
        }
        key = rotate(key);
    }

    return answer;
}

File: 프로그래머스/lv3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.py
def rotate(key):
    n = len(key)
    result = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            x, y =  j, n-i-1
            result[x][y] = key[i][j]
    return result

def check_open(key, lock, n, m, count):
    for x in range(m + n-1):
        for y in range(m + n-1):
            OK = True
            result = 0

            #n의 범위 안에서 하되, 자물쇠에 포함되지 않는 부분은 반복문을 실행하지 않는다.
            for i in range(n):
                for j in range(n):
                    hx = x-(n-1-i)
                    hy = y-(n-1-j)
                    if hx < 0 or hy < 0 or hx >=m or hy >= m: continue
                    if key[i][j] == 1 and lock[hx][hy] == 1:
                        OK = False
                        break
                    elif key[i][j] == 1 and lock[hx][hy] == 0:
                        result += 1
                if not OK:
                    break
            if OK and result == count:
                return True
    return False
    
    
def solution(key, lock):
    holl_count = 0
    n, m = len(key), len(lock)
    for l in lock:
        holl_count += l.count(0)

    answer = False
    for i in range(4):
        answer = check_open(key,lock,n,m, holl_count)
        if answer:
            break
        key = rotate(key)
    
    return answer